codeflash/benchmarking/codeflash_trace.py

```diff
@@ -0,0 +1,78 @@
+import functools
+import os
+import dill as pickle
+import time
+from typing import Callable
+
+
+
+
+class CodeflashTrace:
+    """A class that provides both a decorator for tracing function calls
+    and a context manager for managing the tracing data lifecycle.
+    """
+
+    def __init__(self) -> None:
+        self.function_calls_data = []
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        # Cleanup is optional here
+        pass
+
+    def __call__(self, func: Callable) -> Callable:
+        """Use as a decorator to trace function execution.
+
+        Args:
+            func: The function to be decorated
+
+        Returns:
+            The wrapped function
+
+        """
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            # Measure execution time
+            start_time = time.thread_time_ns()
```

Comment:
if the function is being called in a benchmark more than N number of times then just skip pickling the outputs, but still calculate the timings for your calculations